src/audio/volume/volume_hifi3.c

```diff
@@ -90,6 +112,122 @@ static void vol_s24_to_s24_s32(struct comp_dev *dev, struct audio_stream *sink,
+	ae_f32x2 out_sample;
+	ae_f32x2 volume;
+	size_t channel;
+	/** dsfsefasf */
+	int i;
+	int shift = 8;
+	ae_int32 *in = (ae_int32 *)source->r_ptr;
+	ae_int32 *out = (ae_int32 *)sink->w_ptr;
+	const int inc = sizeof(ae_int32)*sink->channels;
+	/* Processing per channel */
+	for (channel = 0; channel < sink->channels; channel++) {
+
+		/* Load volume */
+		volume = (ae_f32x2)cd->volume[channel];
+
+		/* setting start address of sample load */
+		ae_int32 *in = (ae_int32 *)source->r_ptr + channel;
+
+		/* setting start address of sample store */
+		ae_int32 *out = (ae_int32 *)sink->w_ptr + channel;
+		/* Main processing loop */
+		for (i = 0; i < frames; i++) {
+			/* Set source as circular buffer */
+			vol_setup_circular(source);
+
+			/* Load the input sample */
+			AE_L32_XC(in_sample, in, inc);
+
+			/* Multiply the input sample */
+			mult = AE_MULF32S_LL(volume, AE_SLAA32(in_sample, 8));
+
+			/* Multiplication of Q1.31 x Q1.31 gives Q1.63.
+			 * Now multiplication is Q8.16 x Q1.31, the result
+			 * is Q9.48. Need to shift right by one to get Q17.47
+			 * compatible format for round.
+			 */
+			out_sample = AE_ROUND32F48SSYM(AE_SRAI64(mult, 1));
+
+			/* Shift for S24_LE */
+			out_sample = AE_SRAA32RS(out_sample, shift);
+			out_sample = AE_SLAA32S(out_sample, shift);
+			out_sample = AE_SRAA32(out_sample, shift);
+
+			/* Set sink as circular buffer */
+			vol_setup_circular(sink);
+
+			/* Store the output sample */
+			AE_S32_L_XC(out_sample, out, inc);
+		}
+	}
+#elif (OPT == OPT2)
+	struct vol_data *cd = comp_get_drvdata(dev);
+	ae_f32x2 in_sample = AE_ZERO32();
+	ae_f32x2 out_sample;
+	ae_f32x2 volume;
+	int i;
+	int shift = 8;
+	int channel_num = sink->channels;
+	const int inc = sizeof(ae_f32x2);
+
+	/* store volume twice for 2-way operations */
+	for (i = 0; i < channel_num; i++) {
+		cd->vol[i] = cd->volume[i];
+		cd->vol[i+channel_num] = cd->volume[i];
+	}
+	ae_f32x2 *buf = (ae_f32x2 *)cd->vol;
+	ae_f32x2 *buf_end = (ae_f32x2 *)(cd->vol + channel_num * 2);
+	/* set start address of sample load & store*/
+	ae_f32x2 *in = (ae_f32x2 *)source->r_ptr;
+	ae_f32x2 *out = (ae_f32x2 *)sink->w_ptr;
+	ae_f32x2 *vol = (ae_f32x2 *)buf;
+
+	/* use alignment register to prime the memory to
+	 * avoid risk of buf not aligned to 64 bits.
+	 */
+
+	ae_valign inu, outu;
+
+	AE_LA32X2POS_PC(inu, in);
+	AE_SA64POS_FC(outu, out);
+
+	/* process two continuous sample data once */
+	for (i = 0; i < channel_num * frames; i = i + 2) {
+		/* Set buf who stores the volume gain data as circular buffer */
+		AE_SETCBEGIN0(buf);
+		AE_SETCEND0(buf_end);
+
+		/* Load the volume value */
+		AE_L32X2_XC(volume, vol, inc);
+
+		/* Set source as circular buffer */
+		vol_setup_circular(source);
+
+		/* Load the input sample */
+		AE_LA32X2_IC(in_sample, inu, in);
+
+		/* Multiply the input sample */
+		out_sample = AE_MULFP32X2RS(AE_SLAA32S(volume, 7), AE_SLAA32S(in_sample, 8));
+
+
+		/* Shift for S24_LE */
+		out_sample = AE_SLAA32S(out_sample, shift);
```

Comment:
Shift by immediate 8 could make this faster. There's no more format conversion needed in volume component so the shift variable related code can be all removed.